add aesni crypto implementation

aaronbojarski opened the pull request.

Commit:
add aesni crypto implementation

Commit:
format code

fstreun commented:
Hi Aaron
Thank you for adding the fast crypto implementation to the open-source repo.

Could you please update the 'test.sh' so that we test both versions. Using aesni crypto by default and openssl crypto for one simple test should be sufficient.

Also the documentation and the README.md should be updated accordingly.

Commit:
add integration test with OPENSSL

Commit:
set OPENSSL as default

Commit:
add documentation

Commit:
correct file name

aaronbojarski commented:
Hi Fabio

Thank you for your comments. I have now updated the tests and added a bit of documentation.
While discussing with Marc we decided to make OPENSSL the default and offer AESNI as an optimization.

Let me know if you have any further remarks.

fstreun commented on src/CMakeOptions.cmake:
```diff
@@ -34,8 +34,11 @@ endif()
 
-set(LF_CBCMAC "OPENSSL" CACHE STRING "CBC MAC implementation (OPENSSL).")
-if(LF_CBCMAC MATCHES "OPENSSL")
+set(LF_CBCMAC "OPENSSL" CACHE STRING "CBC MAC implementation (AESNI, OPENSSL).")
+if(LF_CBCMAC STREQUAL "AESNI")
```
Have default as first value in the description. I think that's pretty much the convention here.

```suggestion
set(LF_CBCMAC "OPENSSL" CACHE STRING "CBC MAC implementation (OPENSSL, AESNI).")
if(LF_CBCMAC STREQUAL "AESNI")
```

aaronbojarski replied:
Done

fstreun commented on docs/Installation.md:
```diff
@@ -41,6 +41,7 @@ The following presents an incomplete list:
 LF_WORKER={SCION,IPV4,FWD}
 LF_DRKEY_FETCHER={SCION,MOCK}
 LF_LOG_DP_LEVEL={MIN,MAX,EMERG,ALERT,CRIT,ERR,WARNING,NOTICE,INFO,DEBUG}
+LF_CBCMAC={OPENSSL,AESNI} //NOTE that AESNI is only available on x86 systems with AES assembly instructions. If available AESNI is more performant.
```
A bit too much text for this list. You could add a shorter version of this remark to the description of the option in the CMake file.

```suggestion
LF_CBCMAC={OPENSSL,AESNI}
```

aaronbojarski replied:
Done

fstreun commented on README.md:
Remove the remark regarding the worse performance of the open-source version as this will not be the case anymore with the AESNI crypto.

aaronbojarski replied:
Done

fstreun commented:
> While discussing with Marc we decided to make OPENSSL the default and offer AESNI as an optimization.

Any specific reason for this decision?
I imagine that most deployments would use AESNI.

marcfrei commented:
> I imagine that most deployments would use AESNI.

Agreed.

The idea behind the new default was to optimize for a good first build/run experience also on platforms that don't support AES-NI.

That beeing said, I don't feel to strongly either way. Do you want to decide, @fstreun?

fstreun commented:
> The idea behind the new default was to optimize for a good first build/run experience also on platforms that don't support AES-NI.
> 
> That beeing said, I don't feel to strongly either way. Do you want to decide, @fstreun?

I think that is a reasonable argument. Let's go with an easy first build and use OpenSSL by default. But we should have a section in the documentation that describes how to achieve the best performance. (We can open an issue for that and handle it later)

aaronbojarski committed:
Update src/CMakeOptions.cmake

Co-authored-by: fstreun <fabiostr@bluewin.ch>

Commit:
add some improvements

fstreun commented on src/CMakeOptions.cmake:
```diff
@@ -34,7 +34,7 @@ endif()
 add_compile_definitions(LF_DRKEY_FETCHER_${LF_DRKEY_FETCHER}=1)
 
-set(LF_CBCMAC "OPENSSL" CACHE STRING "CBC MAC implementation (OPENSSL, AESNI).")
+set(LF_CBCMAC "OPENSSL" CACHE STRING "CBC MAC implementation (OPENSSL, AESNI). AESNI available on x86 machines as performance optimization.")
```
`AESNI requires hardware support`

aaronbojarski replied:
Done

Commit:
add recommended comment
